src/Raven.Server/Documents/Queries/Graph/GraphQueryPlan.cs

```diff
@@ -0,0 +1,98 @@
+        private IndexQueryServerSide _query;
+        private DocumentsOperationContext _context;
+        private long? _resultEtag;
+        private OperationCancelToken _token;
+        private DocumentDatabase _database;
+        private IntermediateResults _intermediateResults;
+        public GraphQuery GraphQuery => _query.Metadata.Query.GraphQuery;
+
+        public GraphQueryPlan(IndexQueryServerSide query, DocumentsOperationContext context, long? resultEtag,
+            OperationCancelToken token, DocumentDatabase database)
+        {
+            _database = database;
+            _query = query;
+            _context = context;
+            _resultEtag = resultEtag;
+            _token = token;
+        }
+
+
+        public void BuildQueryPlan()
+        {
+            switch (_query.Metadata.Query.GraphQuery.MatchClause)
+            {
+                case PatternMatchElementExpression pme:
+                    _rootQueryStep = BuildQueryPlanForPattern(pme);
+                    return;
+                default:
+                    throw new ArgumentOutOfRangeException($"Unexpected expression({_query.Metadata.Query.GraphQuery.MatchClause.Type}) type for MatchClause");
+            }
+
+            
+        }
+
+        private IQueryStep BuildQueryPlanForPattern(PatternMatchElementExpression patternExpression)
+        {
+            IQueryStep prev = BuildQueryPlanForMatchVertex(patternExpression.Path[0]);
+            for (var i = 2; i < patternExpression.Path.Length; i+=2)
+            {
+                var tmpVertex = BuildQueryPlanForMatchVertex(patternExpression.Path[i]);
+                prev = BuildQueryPlanForEdge(prev, tmpVertex, patternExpression.Path[i-1]);
+            }
+
+            return prev;
+        }
+
+        private IQueryStep BuildQueryPlanForEdge(IQueryStep left, IQueryStep right, MatchPath edge)
+        {
+            var allias = edge.Alias;
+            if (GraphQuery.WithEdgePredicates.TryGetValue(allias, out var withEdge) == false)
+            {
+                throw new InvalidOperationException($"BuildQueryPlanForEdge was invoked for allias='{allias}' which suppose to be an edge but no corresponding WITH EDGE clause was found.");
+            }
+
+            return new EdgeQueryStep(left, right, withEdge, edge, _query.QueryParameters);
+        }
+
+        private IQueryStep BuildQueryPlanForMatchVertex(MatchPath vertex)
+        {            
+            Sparrow.StringSegment alias = vertex.Alias;
+            if (GraphQuery.WithDocumentQueries.TryGetValue(alias, out var query) == false)
+            {
+                throw new InvalidOperationException($"BuildQueryPlanForMatchVertex was invoked for allias='{alias}' which suppose to be a vertex but no corresponding WITH clause was found.");
+            }
+            //TODO: we can tell at this point if it is a collection query or not, we might want to build a diffrent step for collection queries in the future.        
```

Comment:
We want to have:

* Index step
* Collection source step
* Collection dest step
* Collection dest step + filter

Reply:
The question is would we want to go complete OOP and have those as subclasses or would an enum be enough since they will all require pretty much the same parameters and logic to run.